### PR TITLE
chore(reth_primitives): Use trait for `size` methods in primitive types

### LINKED ISSUE #3893: Use a trait for `size` methods in primitive types
### Describe the feature

Many primitive types have methods which calculate a heuristic for the heap size of the type, for example:
https://github.com/paradigmxyz/reth/blob/b69a18dc47255304b3943ca859dc6eb9da579b19/crates/primitives/src/block.rs#L192-L200

This should be captured in a trait, and it should be possible to `derive` this trait for all of these types.

One crate that is interesting for this is [DeepsizeOf](https://docs.rs/deepsize/latest/deepsize/trait.DeepSizeOf.html). An unanswered question is whether or not we should use this crate, because its implementation of `DeepSizeOf` for `Vec` seems to contradict the rust docs.

The implementation:
```rust
impl<T> DeepSizeOf for std::vec::Vec<T> where T: DeepSizeOf {
    fn deep_size_of_children(&self, context: &mut Context) -> usize {
        // Size of heap allocations for each child
        self.iter().map(|child| child.deep_size_of_children(context)).sum()
         + self.capacity() * std::mem::size_of::<T>()  // Size of Vec's heap allocation
    }
}
```
The [**Guarantees** section of the rust docs](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#guarantees) state:

> However, the pointer might not actually point to allocated memory. In particular, if you construct a `Vec` with capacity 0 via [`Vec::new`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.new), [`vec![]`](https://doc.rust-lang.org/stable/std/macro.vec.html), [`Vec::with_capacity(0)`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.with_capacity), or by calling [`shrink_to_fit`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.shrink_to_fit) on an empty Vec, it will not allocate memory. Similarly, if you store zero-sized types inside a `Vec`, it will not allocate space for them. Note that in this case the `Vec` might not report a [`capacity`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.capacity) of 0. `Vec` will allocate if and only if [`mem::size_of::<T>()`](https://doc.rust-lang.org/stable/std/mem/fn.size_of.html)` * `[`capacity()`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.capacity)` > 0`. In general, Vec’s allocation details are very subtle — if you intend to allocate memory using a `Vec` and use it for something else (either to pass to unsafe code, or to build your own memory-backed collection), be sure to deallocate this memory by using `from_raw_parts` to recover the `Vec` and then dropping it.
> 
> If a `Vec` has allocated memory, then the memory it points to is on the heap (as defined by the allocator Rust is configured to use by default), and its pointer points to [`len`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.len) initialized, contiguous elements in order (what you would see if you coerced it to a slice), followed by [`capacity`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.capacity)` - `[`len`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.len) logically uninitialized, contiguous elements.

This leads me to believe that the correct implementation should subtract `len` from `capacity` before multiplying by `std::mem::size_of::<T>()`. If this is true, then the crate may slightly overestimate the memory used by a `Vec<T>`.

Ultimately, this is not a huge problem, and using `DeepSizeOf` would be the _easiest and most straightforward_ way to derive this size heuristic, and we could upstream a fix if required. Our implementation also does not subtract the `len` from `capacity`, so using the `DeepSizeOf` crate should also match our existing logic.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TropicalDog17**
Closes #3893

### CODE REVIEW

File: crates/primitives-traits/src/size.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/// Trait for calculating a heuristic for the in-memory size of a struct.
+pub trait HeuristicSize {
```
Comment (Rjected) [inline crates/primitives-traits/src/size.rs:2]: maybe name it `SizeHeuristic`, or `InMemorySize` or something like that

File: crates/prune/prune/src/segments/user/receipts_by_logs.rs [hunk 2]
```diff
@@ -4,13 +4,14 @@ use crate::{
     PrunerError,
 };
 use reth_db::{tables, transaction::DbTxMut};
+#[allow(unused_imports)]
```
Comment (Rjected) [inline crates/prune/prune/src/segments/user/receipts_by_logs.rs:7]: should remove this

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs [hunk 3]
```diff
@@ -160,3 +160,14 @@ pub fn transaction_to_call_request(tx: TransactionSignedEcRecovered) -> Transact
...
+/// Convert [`reth_primitives::TxType`] to [`alloy_consensus::TxType`]
+/// Will panics if the `TxType` is `TxType::Deposit` (Optimism Deposit transaction)
+pub fn tx_type_to_alloy(tx_type: TxType) -> alloy_consensus::TxType {
+    u8::from(tx_type).try_into().unwrap()
+}
+
+/// Convert [`alloy_consensus::TxType`] to [`reth_primitives::TxType`]
+pub fn tx_type_from_alloy(tx_type: alloy_consensus::TxType) -> TxType {
+    u8::from(tx_type).try_into().unwrap()
+}
```
Comment (Rjected) [inline crates/rpc/rpc-types-compat/src/transaction/mod.rs:173]: these changes are unrelated right?

File: crates/transaction-pool/src/lib.rs [hunk 4]
```diff
@@ -157,6 +157,7 @@ use aquamarine as _;
 use reth_eth_wire_types::HandleMempoolData;
 use reth_execution_types::ChangedAccount;
 use reth_primitives::{BlobTransactionSidecar, PooledTransactionsElement};
+use reth_primitives_traits as _;
```
Comment (Rjected) [inline crates/transaction-pool/src/lib.rs:160]: should remove this

File: crates/transaction-pool/src/test_utils/mock.rs [hunk 5]
```diff
@@ -23,9 +23,11 @@ use reth_primitives::{
     BlobTransactionValidationError, PooledTransactionsElementEcRecovered, Signature, Transaction,
     TransactionSigned, TransactionSignedEcRecovered, TxType,
 };
-
 use std::{ops::Range, sync::Arc, time::Instant, vec::IntoIter};
 
+#[cfg(any(test, feature = "arbitrary"))]
+use reth_primitives_traits::HeuristicSize;
```
Comment (Rjected) [inline crates/transaction-pool/src/test_utils/mock.rs:29]: if the `use reth_primitives_traits as _;` is required since this is feature gated, I'd try to instead make primitives-traits an optional dep, including it by default as a dev-dependency, and when `arbitrary` is enabled

**Review by Rjected — DISMISSED** [review]
great start, I have a few suggestions

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs (hunk 3 above)
Comment (TropicalDog17) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs:173]: yeah sorry I messed up my local branch

**Review by mattsse — APPROVED** [review]
seems okay, pending @Rjected
